internal/testprocessor/testscript.go

```diff
@@ -0,0 +1,128 @@
+package testprocessor
+
+import (
+	"encoding/csv"
+	"fmt"
+	"os"
+	"time"
+
+	zlog "github.com/rs/zerolog/log"
+)
+
+type Mode string
+
+const (
+	SerialMode              Mode = "serial"
+	SerialWithoutDelaysMode Mode = "serial-without-delays"
+	ParallelMode            Mode = "parallel"
+)
+
+type Config struct {
+	Mode         Mode
+	RunsDataPath string
+	OutputPath   string
+	DefaultQuery *string
+	Percentiles  []int
+}
+
+type PlaygroundClient interface {
+	PostRuns(database string, version string, query string) (time.Duration, error)
+}
+
+type Processor struct {
+	PlaygroundClient PlaygroundClient
+	Config           *Config
+}
+
+func New(playgroundClient PlaygroundClient, config *Config) *Processor {
+	return &Processor{
+		PlaygroundClient: playgroundClient,
+		Config:           config,
+	}
+}
+
+func (p *Processor) Process() {
+	runs, err := loadRuns(p.Config.RunsDataPath)
+	if err != nil {
+		zlog.Fatal().Err(err).Msg("runs data cannot be loaded")
+	}
+
+	if p.Config.DefaultQuery != nil {
+		for _, run := range runs.Runs {
+			run.Query = *p.Config.DefaultQuery
+		}
+	}
+
+	switch p.Config.Mode {
+	case SerialMode:
+		p.processSerialMode(runs)
+	case SerialWithoutDelaysMode:
+		p.processSerialWithoutDelaysMode(runs)
+	case ParallelMode:
+		p.processParallelMode()
+	default:
+		zlog.Fatal().Msg("unknown mode")
+	}
+
+	p.exportTestResult(runs.Runs)
+
+	aggregator := NewAggregator(runs.Runs)
+	aggregator.PrintPercentiles(p.Config.Percentiles)
+}
+
+func (p *Processor) processSerialMode(runs *RunsInput) {
+	for i, run := range runs.Runs {
+		runResult, err := p.PlaygroundClient.PostRuns(run.Database, run.Version, run.Query)
+		if err != nil {
+			zlog.Error().Err(err).Msg("received error from playground client")
+		}
+		run.TimeElapsed = runResult
+		zlog.Info().Msg(fmt.Sprintf("processed request: %s", runResult.String()))
+
+		if i != len(runs.Runs)-1 {
+			nextRequestTime := runs.Runs[i+1].Timestamp
+			sleepDelta := nextRequestTime.Sub(run.Timestamp)
+			time.Sleep(sleepDelta)
+		}
+	}
+}
+
+func (p *Processor) processSerialWithoutDelaysMode(runs *RunsInput) {
+	for _, run := range runs.Runs {
+		runResult, err := p.PlaygroundClient.PostRuns(run.Database, run.Version, run.Query)
+		if err != nil {
+			zlog.Error().Err(err).Msg("received error from playground client")
+		}
+		run.TimeElapsed = runResult
+		zlog.Info().Msg(fmt.Sprintf("Processed request: %s", runResult.String()))
```

Comment:
You can use log fields instead of fmt format strings.

Besides, you can use fmt.Print*** to better control the output of your application (loggers usually are good for collecting logs via applications and viewing them in other UI).